### PR TITLE
Only write the pid-file while listening to SIGUSR signals.

### LINKED ISSUE #6970: flutter run: SIGUSR1 not always caught by Dart process
Sometimes, when I have `flutter run` running, and I send a SIGUSR1 to the process, I see:

```
~/dev/flutter/bin/flutter: line 53: 14926 User defined signal 1   "$DART" "$SNAPSHOT_PATH" "$@"
```

cc @johnmccutchan

### PR DESCRIPTION
**Pull request by Hixie**
This prevents race conditions where you send the signal early and kill the tool.

This also introduces some tests for vaguely-related features like hot reload and such. It reuses a test file that was previously used by a devicelab test but that test was removed leaving this file as unused.

Fixes https://github.com/flutter/flutter/issues/6970.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by Hixie** [comment]
blocked on https://github.com/flutter/flutter/pull/74859

**Comment by Hixie** [comment]
Once #74859 lands I need to add a test for the DevTools error messages.

**Comment by Hixie** [comment]
cc @jonahwilliams

File: packages/flutter_tools/test/integration.shard/overall_experience_test.dart
Comment (jonahwilliams) [inline packages/flutter_tools/test/integration.shard/overall_experience_test.dart:34]: I don't think you need this ignores if this file is opted out with the 2.8 above

**Review by jonahwilliams — COMMENTED** [review]
Still reviewing - but I really like this.

I'll spend some time this afternoon seeing how I can get these tests to break :)

File: packages/flutter_tools/test/integration.shard/overall_experience_test.dart
Comment (Hixie) [reply packages/flutter_tools/test/integration.shard/overall_experience_test.dart:34]: oh right, i added these before i realised i needed to just opt this file out

**Comment by jonahwilliams** [comment]
I did not end up having time to get to this after crash triage follow ups today. Will take a look tomorrow morning

**Comment by Hixie** [comment]
> I'll spend some time this afternoon seeing how I can get these tests to break :)

Seems you won't have to try hard... for some reason on the bots they don't ever see DevTools starting up. Can't figure out why yet...

**Comment by jonahwilliams** [comment]
I think the messaging could be a bit clearer about the expectation when a transition is missed. For example, I change the tool to print "Application done" instead of "Application finished":

The PID test fails with:
```
Missed some expected transitions.
Expected state transitions:
 DONE  "q Quit (terminate the application on the device)."
 NEXT> "Application finished."
Log (only contains logged lines; use debug:true to print all output):
...
```

I think this means it had matched the DONE line and was looking for the NEXT line right? If that is the case I would put NEXT first since it is more important, and DONE is really just contextual. I think there might be a way to phrase this a little differently to make it clearer that DONE is the current state and not missing?

```
Missed some expected transitions.
Expected state transitions:
 NEXT> "Application finished."
 DONE  "q Quit (terminate the application on the device)."
Log (only contains logged lines; use debug:true to print all output):
...
```

Second test was adding an unhandled exception, in this case the default logging captures the error as expected. I wonder if it would be worth calling out in the expectation message that an unexpected error was thrown?

**Comment by Hixie** [comment]
> I think this means it had matched the DONE line and was looking for the NEXT line right? If that is the case I would put NEXT first since it is more important, and DONE is really just contextual. I think there might be a way to phrase this a little differently to make it clearer that DONE is the current state and not missing?

I'll look into making it clearer, good call. I'm not sure about reversing the order, unless I make the "fail" case use different logic than the "it's taking a long time" case. For example, change things so that it gets stuck half way and it's show a bunch of "DONE" lines followed by the "NEXT" line it's waiting for followed by a bunch of lines that it hasn't done yet, and reversing the order seems like it'd be confusing to me.



> Second test was adding an unhandled exception, in this case the default logging captures the error as expected. I wonder if it would be worth calling out in the expectation message that an unexpected error was thrown?

Not quite sure I follow, can you elaborate?

**Comment by jonahwilliams** [comment]
Sorry, what I mean is if the tool crashes, it might be better to call that out first above the particular matcher state. That is more of a nit, since the stack trace makes it into stdout instead.

**Comment by Hixie** [comment]
(PR has been updated with better diagnostics based on suggestions above.)

**Comment by Hixie** [comment]
> Sorry, what I mean is if the tool crashes, it might be better to call that out first above the particular matcher state. That is more of a nit, since the stack trace makes it into stdout instead.

Gotcha. Any idea how to notice the crash? I report the exit code, but beyond that I'm not sure what to do...

**Comment by Hixie** [comment]
added this:
```
    if (stderr.contains('Oops; flutter has exited unexpectedly:')) {
      print('The subprocess may in fact have crashed. Check the stderr logs below.');
    }
```
...to the "The subprocess terminated before all the expected transitions had been matched." case.

**Comment by Hixie** [comment]
Looks like there's an issue with launching devtools; I'm going to temporarily comment out those tests and work on fixing them in a separate PR so that I can get this test harness in first.

File: packages/flutter_tools/test/integration.shard/overall_experience_test.dart
Comment (jonahwilliams) [inline packages/flutter_tools/test/integration.shard/overall_experience_test.dart:147]: I think it is possible for the test to fail and the next one to start before the process gets killed. I noticed that if writing the PID file fails the expect on L279, then the next test will always fail (on windows) because the previous flutter instance is still going. You might need something like a `finally` block that kills the process if a test exception is thrown

File: packages/flutter_tools/test/integration.shard/overall_experience_test.dart
Comment (Hixie) [reply packages/flutter_tools/test/integration.shard/overall_experience_test.dart:302]: I just reused the same test file that was being used in the test you removed. :-)

File: packages/flutter_tools/test/integration.shard/overall_experience_test.dart
Comment (Hixie) [reply packages/flutter_tools/test/integration.shard/overall_experience_test.dart:361]: `flutter run` will imply a `pub get` so I think that works regardless.

File: packages/flutter_tools/lib/src/base/io.dart
Comment (Hixie) [reply packages/flutter_tools/lib/src/base/io.dart:388]: This was just copy/pasted from _DefaultProcessInfo. I've changed both to `_fileSystem`.

File: packages/flutter_tools/test/integration.shard/overall_experience_test.dart
Comment (Hixie) [reply packages/flutter_tools/test/integration.shard/overall_experience_test.dart:206]: done and done

File: packages/flutter_tools/test/integration.shard/overall_experience_test.dart
Comment (Hixie) [reply packages/flutter_tools/test/integration.shard/overall_experience_test.dart:147]: There's not much I can do about this because the subprocess is a shell process, not the Dart process, so killing it doesn't actually do anything useful.

I've made the first test avoid expects in the callbacks, which addresses the immediate concern you mention. Not much I can do to enforce that though.

That said, I actually don't think you're right. Each test is generating its own temporary directory with its own PID file, so zombie flutter processes from earlier tests shouldn't affect later tests. (Unless you're saying on Windows you can never run `flutter` twice at the same time? In which case, I wasn't aware of that. Still not much I can do though.)

Comment (jonahwilliams) [reply packages/flutter_tools/test/integration.shard/overall_experience_test.dart:147]: I think it's specific to how we `flutter run` on the tester device - we copy the kernel artifacts to a generic location under build/ , but instead it should probably be to a temp directory. Filled https://github.com/flutter/flutter/issues/75534

But at any rate, after the test failed and exited I still have some zombie flutter_tester processes running on windows. Maybe write a `q` into stdin to try and shut them down gracefully?

File: packages/flutter_tools/test/integration.shard/overall_experience_test.dart
Comment (jonahwilliams) [reply packages/flutter_tools/test/integration.shard/overall_experience_test.dart:372]: ahh makes sense

File: packages/flutter_tools/test/integration.shard/overall_experience_test.dart
Comment (Hixie) [reply packages/flutter_tools/test/integration.shard/overall_experience_test.dart:157]: Added this:
```patch
-  final int exitCode = await process.exitCode;
+  final int exitCode = await process.exitCode
+    ..timeout(const Duration(seconds: 15), () {
+      print('(process is not quitting, trying to send a "q" just in case that helps)');
+      print('(a functional test should never reach this point)');
+      process.stdin.write('q');
+    });
```

**Review by jonahwilliams — APPROVED** [review]
I really like the new test harness, looking forward to putting it to good use.


LGTM (but not to CI)

File: packages/flutter_tools/test/integration.shard/overall_experience_test.dart
Comment (Hixie) [reply packages/flutter_tools/test/integration.shard/overall_experience_test.dart:157]: ...I swear I tested that, but somehow the code above (and the code I actually landed) was bogus. Anyway, I've fixed that now...

**Comment by Hixie** [comment]
Thanks! Will land on green.

**Pull request merged**
